lgc/patch/VertexFetch.cpp

```diff
@@ -1035,9 +802,183 @@ Value *VertexFetchImpl::fetchVertex(InputImportGenericOp *inst, llvm::Value *des
   const unsigned bitWidth = basicTy->getScalarSizeInBits();
   assert(bitWidth == 8 || bitWidth == 16 || bitWidth == 32 || bitWidth == 64);
 
-  auto func = generateFetchFunction(bitWidth, inst->getFunction()->getParent());
-  Value *lastVert =
-      builder.CreateCall(func, {vbDesc, vbIndex, byteOffset, componentSize, isPacked, isBgr, yMask, zMask, wMask});
+  // If ispacked is false, we require per-component fetch
+  builder.CreateCondBr(isPacked, wholeVertexBlock, comp0Block);
+
+  // 8-bit vertex fetches use the 16-bit path as well.
+  bool is16bitFetch = bitWidth <= 16;
+  bool is64bitFetch = bitWidth == 64;
+  auto fetch64Type = FixedVectorType::get(Type::getInt32Ty(*m_context), 2);
+  auto fetch32Type = FixedVectorType::get(Type::getInt32Ty(*m_context), 4);
+  auto fetch16Type = FixedVectorType::get(Type::getInt16Ty(*m_context), 4);
+  auto fetchType = is16bitFetch ? fetch16Type : fetch32Type;
+
+  Value *args[] = {
+      vbDesc,              // rsrc
+      vbIndex,             // vindex
+      byteOffset,          // offset
+      builder.getInt32(0), // soffset
+      builder.getInt32(0)  // glc, slc
+  };
+
+  Value *wholeVertex = nullptr;
+  {
+    builder.SetInsertPoint(wholeVertexBlock);
+    wholeVertex = builder.CreateIntrinsic(Intrinsic::amdgcn_struct_buffer_load_format, fetchType, args, {});
+    if (is64bitFetch) {
+      // If it is 64-bit, we need the second fetch
+      args[2] = builder.CreateAdd(args[2], builder.getInt32(SizeOfVec4));
+      auto secondFetch = builder.CreateIntrinsic(Intrinsic::amdgcn_struct_buffer_load_format, fetchType, args, {});
+      std::vector<Constant *> shuffleMask;
+      for (unsigned i = 0; i < 8; ++i)
+        shuffleMask.push_back(ConstantInt::get(Type::getInt32Ty(*m_context), i));
+      wholeVertex = builder.CreateShuffleVector(wholeVertex, secondFetch, ConstantVector::get(shuffleMask));
+    }
+    builder.CreateBr(fetchEndBlock);
+  }
+
+  fetchType = FixedVectorType::get(builder.getInt32Ty(), is64bitFetch ? 8 : 4);
+  if (is16bitFetch)
+    fetchType = FixedVectorType::get(builder.getInt16Ty(), 4);
+  Value *lastVert = UndefValue::get(fetchType);
+
+  auto int16Zero = builder.getInt16(0);
+  auto int16One = builder.getInt16(1);
+  auto float16One = builder.getInt16(0x3C00);
+  // Get default fetch values
+  Constant *defaults = nullptr;
+  if (basicTy->isIntegerTy()) {
+    if (bitWidth <= 16)
+      defaults = ConstantVector::get({int16Zero, int16Zero, int16Zero, int16One});
+    else if (bitWidth == 32)
+      defaults = m_fetchDefaults.int32;
+    else {
+      assert(bitWidth == 64);
+      defaults = m_fetchDefaults.int64;
+    }
+  } else if (basicTy->isFloatingPointTy()) {
+    if (bitWidth == 16)
+      defaults = ConstantVector::get({int16Zero, int16Zero, int16Zero, float16One});
+    else if (bitWidth == 32)
+      defaults = m_fetchDefaults.float32;
+    else {
+      assert(bitWidth == 64);
+      defaults = m_fetchDefaults.double64;
+    }
```

Comment:
For consistency, use member variables `m_fetchDefaults.int16` and `m_fetchDefaults.float16`

Reply:
The expected type is `Int16Ty` here, but `m_fetchDefaults.int16` is `Int32Ty`.

Reply:
In `addVertexFetchInst`, after calling `llvm.amdgcn.struct.tbuffer.load`, extend 16-bit to 32-bit. Keeping 16-bit may be a better choice. I will add a commit to fix it.